### PR TITLE
fix(train): crash-safe checkpoints — completeness marker gating resume

### PR DESCRIPTION
**Pull request by jessiewei7**
# fix(train): crash-safe checkpoints — completeness marker gating resume

## Summary

`save_checkpoint()` writes `model.safetensors` / `optimizer_state_dict.pt` /
etc. directly to their final paths, and `_get_previous_epoch()` (which resume
uses to find the latest checkpoint) trusts **any directory whose name parses
as an integer** — nothing distinguishes a fully written checkpoint from one
truncated by a crash mid-save.

Failure mode:

1. `save_checkpoint()` for epoch N is killed mid-`save_pretrained()` (OOM,
   preemption, node failure) — directory `N/` now exists with a truncated
   `model.safetensors`.
2. The job restarts; `_get_previous_epoch()` sees `N`, returns it.
3. `load_model_state_dict()` hits the truncated file: best case safetensors
   throws a parse error and the run is stuck until someone deletes `N/` by
   hand; worst case (`optimizer_state_dict.pt` via `torch.load`) a partial
   read surfaces as a confusing unpickling error.

`training_state.json` does not guard this: resume only reads it to pick
mid-epoch vs end-of-epoch, and a missing/corrupt one silently falls back to
"advance to next epoch" — still loading model/optimizer from the broken dir.

## Fix

SpecForge had the identical bug class; the fix follows the same design as
sgl-project/SpecForge#659: a completeness marker written last, gating resume.

- `BaseCheckpointer.mark_checkpoint_complete(epoch)`: fsync every file in the
  epoch dir, then write a `checkpoint_complete` marker (fsync it and the
  directory). The fsyncs matter: after a hard host crash, files can survive
  at full length with truncated content, so the marker must not claim
  completeness before the data is durable.
- `Trainer` calls it once all files of an epoch are in place (model,
  optimizer, scheduler, `training_state.json`, val metrics — the last writes
  happen in the trainer, which is why the marker call lives there).
- `_get_previous_epoch()` skips integer dirs without the marker, logging a
  warning with explicit instructions (`touch .../checkpoint_complete`) —
  the same warn-and-instruct pattern already used for `interrupted/` dirs.
  This also covers checkpoints from before this change.

## Test plan

`tests/unit/train/test_checkpoint.py` (extends the existing suite, CPU-only):

- a dir with files but no marker is skipped; resume falls back to the newest
  complete checkpoint (the crash scenario);
- `mark_checkpoint_complete` makes a checkpoint visible to a fresh
  checkpointer;
- `maybe_save_checkpoint` end-to-end writes the marker.

Full `tests/unit/train/` suite passes (113 passed; 1 pre-existing failure is
a hub-connectivity issue in `test_setup_model.py`, unrelated).

Also require `checkpoint_best` to resolve to a completed checkpoint, so best-validation state cannot be restored from an incomplete epoch.

### CODE REVIEW

File: src/speculators/train/checkpointer.py [hunk 1]
```diff
@@ -116,9 +117,19 @@ def _get_previous_epoch(self) -> int:
...
+                if not (d / self.COMPLETE_MARKER_FILENAME).exists():
+                    logger.warning(
+                        f"Skipping checkpoint at {d}: missing "
+                        f"'{self.COMPLETE_MARKER_FILENAME}' marker (save was "
+                        "interrupted, or the checkpoint predates this version). "
+                        "To resume from it anyway, create the marker: "
+                        f"'touch {d / self.COMPLETE_MARKER_FILENAME}'."
+                    )
+                    continue
+                last_checkpoint_num = max(last_checkpoint_num, epoch_num)
```
[the remark this answers is not in the document]
Comment (jessiewei7) [reply src/speculators/train/checkpointer.py:151]: Good catch — mid-epoch saves rewrite the same epoch directory in place, so the previous save's marker could survive a crash during the rewrite. Fixed in ad52938: the trainer now removes the stale marker (`clear_checkpoint_complete`) before each re-save, and the marker is only recreated after the full save completes. Added a regression test asserting the marker is absent while the rewrite is in flight.

**Review by orestis-z — APPROVED** [review]
LGTM — reviewed design, fsync ordering, distributed rank-gating, and test coverage. The completeness-marker pattern is sound, the stale-marker edge case is handled, and the regression test covers the core crash scenario.

**Comment by jessiewei7** [comment]
Rebased onto latest main. CI needs maintainer approval to run whenever you have a moment — thanks!

**Comment by fynnsu** [comment]
@jessiewei7 looks like a test is failing: https://buildkite.com/vllm/speculators/builds/906/list?jid=019f5d27-9efb-43a5-83c5-0f0a8d09b39b&tab=output

Could you take a look?

**Comment by jessiewei7** [comment]
Fixed — test fixtures called `save_checkpoint()` directly, skipping `mark_checkpoint_complete()`.
